### ISSUE TITLE
False positive for `redundant_nil_coalescing` for double Optional.

### ISSUE THREAD

**Issue by hesyifei** [realm/SwiftLint#4738]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

False positive for `redundant_nil_coalescing` check when we have double Optional.

Consider the following code:
```swift
import Foundation

func myFirstFunc() -> String?? {
    nil
}
func mySecondFunc() -> String? {
    return myFirstFunc() ?? nil
}
```

Note that this `?? nil` is not redundant because of the double Optional returned by the first function. So this is a false positive for `redundant_nil_coalescing`.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ ./swiftlint example.swift
Linting Swift files at paths example.swift
Linting 'example.swift' (1/1)
example.swift:7:26: warning: Redundant Nil Coalescing Violation: nil coalescing operator is only evaluated if the lhs is nil, coalescing operator with nil as rhs is redundant (redundant_nil_coalescing)
Done linting! Found 1 violation, 0 serious in 1 file.
```

Furthermore, when we run
```bash
$ ./swiftlint --fix example.swift
Correcting Swift files at paths example.swift
Correcting 'example.swift' (1/1)
example.swift:7:25 Corrected Redundant Nil Coalescing
Done correcting 1 file!
```
, it removes the `?? nil` from the second function, resulting in
```swift
func mySecondFunc() -> String? {
    return myFirstFunc()
}
```
which actually makes the code fail to compile with error
```
error: value of optional type 'String??' must be unwrapped to a value of type 'String?'
 return myFirstFunc()
 ^
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
  * 0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
  * `portable_swiftlint` from https://github.com/realm/SwiftLint/releases/download/0.50.3/portable_swiftlint.zip
* Paste your configuration file:

```yml
opt_in_rules:
  - redundant_nil_coalescing
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.
  * No
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
import Foundation

func myFirstFunc() -> String?? {
    nil
}
func mySecondFunc() -> String? {
    // This triggers a violation:
    return myFirstFunc() ?? nil
}
```

**Comment by SimplyDanny** [comment]
As this rule works on the plain source-based AST without type information it's not possible to fix this false-positive. A workaround would be to use `?? .none` instead of `?? nil` at the moment. But actually this is rather a bug than a feature as they are semantically equivalent.

The only option is to implement another version of this rule as an Analyzer rule that is much more expensive. I'll keep this issue open as an enhancement for that reason.
